amm/contracts/stable_pool/lib.rs

```diff
@@ -0,0 +1,1033 @@
+                return Err(StablePoolError::InsufficientOutputAmount);
+            };
+            // update reserves
+            self.increase_reserve(token_in_id, token_in_amount)?;
+            self.decrease_reserve(token_out_id, token_out_amount)?;
+
+            // mint admin fee
+            self.mint_admin_fee(fee, token_out_id)?;
+            Ok((token_out_amount, fee))
+        }
+
+        /// This method is for internal use only
+        /// - calculates token_in amount
+        /// - calculates swap fee
+        /// - mints admin fee
+        /// - updates reserves
+        /// It assumes that rates have been updated.
+        /// Returns (token_in_amount, swap_fee)
+        fn _swap_from(
+            &mut self,
+            token_in_id: usize,
+            token_out_id: usize,
+            token_out_amount: u128,
+            max_token_in_amount: u128,
+        ) -> Result<(u128, u128), StablePoolError> {
+            if token_out_amount == 0 {
+                return Err(StablePoolError::InsufficientOutputAmount);
+            }
+            let rates = self.get_scaled_rates()?;
+            // calc amount_out and fees
+            let (token_in_amount, fee) = math::rated_swap_from(
+                &rates,
+                token_in_id as usize,
+                token_out_amount,
+                token_out_id as usize,
+                &self.reserves(),
+                &self.pool.fees,
+                self.amp_coef(),
+            )?;
+
+            // Check if in token_in_amount is as constrained by the user
+            if token_in_amount > max_token_in_amount {
+                return Err(StablePoolError::TooLargeInputAmount);
+            };
+            // update reserves
+            self.increase_reserve(token_in_id, token_in_amount)?;
+            self.decrease_reserve(token_out_id, token_out_amount)?;
+
+            // mint admin fee
+            self.mint_admin_fee(fee, token_out_id)?;
+
+            // note that fee is applied to token_out (same as in _swap_to)
+            Ok((token_out_amount, fee))
+        }
+    }
+
+    impl StablePool for StablePoolContract {
+        #[ink(message)]
+        fn add_liquidity(
+            &mut self,
+            min_share_amount: u128,
+            amounts: Vec<u128>,
+            to: AccountId,
+        ) -> Result<(u128, u128), StablePoolError> {
+            // Make sure rates are up to date before we attempt any calculations
+            self.update_rates();
+            if amounts.len() != self.pool.tokens.len() {
+                return Err(StablePoolError::IncorrectAmountsCount);
+            }
+
+            // transfer amounts
+            for (id, &token) in self.pool.tokens.iter().enumerate() {
+                self.token_by_address(token).transfer_from(
+                    self.env().caller(),
+                    self.env().account_id(),
+                    amounts[id],
+                    vec![],
+                )?;
+            }
```

Comment:
Why not assume tokens have been transferred? That's how it's done in UniV2 and I think this design has advantages.

Reply:
It will make writing router simpler.